File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head==null){
            return null;
        }
        head = connectBw(head);
        connectRandom(head);
        return deepCopy(head);
    }
    public Node connectBw(Node head){
        Node temp = head;
        while(temp!=null){
            Node copyNode = new Node(temp.val);
            copyNode.next = temp.next;
            temp.next = copyNode;
            temp = temp.next.next;
        }
        return head;
    }
    public void connectRandom(Node head){
        Node temp = head;
        while(temp!=null){
            Node copyNode = temp.next;
            if(temp.random!=null){
                copyNode.random = temp.random.next;
            }else{
                copyNode.random = null;
            }
            temp = temp.next.next;

        }
    }
    public Node deepCopy(Node head){
        //in this function we connect our copied LL 
        Node dummyNode = new Node(-1);
        Node res = dummyNode;
        Node temp = head;
        while (temp!=null){
            //connecting our copied ll to dummy node
            res.next = temp.next;
            temp.next = temp.next.next;
            //moving to next connection
            res = res.next;
            temp = temp.next;
        }
        return dummyNode.next;
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
class Node:
    def __init__(self, val=0, next=None, random=None):
        self.val = val
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        
        head = self.connectBw(head)
        self.connectRandom(head)
        return self.deepCopy(head)

    def connectBw(self, head: 'Node') -> 'Node':
        temp = head
        while temp:
            copyNode = Node(temp.val)
            copyNode.next = temp.next
            temp.next = copyNode
            temp = temp.next.next
        return head

    def connectRandom(self, head: 'Node'):
        temp = head
        while temp:
            copyNode = temp.next
            if temp.random:
                copyNode.random = temp.random.next
            else:
                copyNode.random = None
            temp = temp.next.next

    def deepCopy(self, head: 'Node') -> 'Node':
        # Separate the copied list from the original list
        dummyNode = Node(-1)
        res = dummyNode
        temp = head
        while temp:
            # Connect the copied list to the dummy node
            res.next = temp.next
            temp.next = temp.next.next
            # Move to the next connection
            res = res.next
            temp = temp.next
        return dummyNode.next
